src/AuditTrailModule/javasource/test_crm/tests/expected/ExpectedLogLineReferenceBase.java

```diff
@@ -0,0 +1,68 @@
+package test_crm.tests.expected;
+
+import static org.junit.Assert.*;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import com.mendix.core.CoreException;
+import com.mendix.systemwideinterfaces.core.IContext;
+import com.mendix.systemwideinterfaces.core.IMendixObject;
+
+import audittrail.proxies.LogLine;
+import audittrail.proxies.TypeOfReferenceLog;
+import test_crm.tests.actual.ActualLogLine;
+import test_crm.tests.actual.ActualReferenceLog;
+
+public abstract class ExpectedLogLineReferenceBase extends ExpectedLogLine {
+	final IContext context;
+	final private Map<Long, ExpectedReferenceLog> referenceChanges = new HashMap<Long, ExpectedReferenceLog>();
+
+	public ExpectedLogLineReferenceBase(final IContext context, final String memberName) {
+		super(memberName);
+		this.context = context;
+	}
+
+	public ExpectedLogLineReferenceBase add(final IMendixObject reference) {
+		this.referenceChanges.put(reference.getId().toLong(), new ExpectedReferenceLog(this.context, reference, TypeOfReferenceLog.Added));
+		return this;
+	}
+
+	public ExpectedLogLineReferenceBase delete(final IMendixObject reference) {
+		this.referenceChanges.put(reference.getId().toLong(), new ExpectedReferenceLog(this.context, reference, TypeOfReferenceLog.Deleted));
+		return this;
+	}
+
+	public ExpectedLogLineReferenceBase keep(final IMendixObject reference) {
+		this.referenceChanges.put(reference.getId().toLong(), new ExpectedReferenceLog(this.context, reference, TypeOfReferenceLog.No_Change));
+		return this;
+	}
+
+
+	public void verify(final ActualLogLine actualLogLine) throws CoreException {
+		super.verify(actualLogLine);
+		final LogLine logLine = actualLogLine.getLogLine();
+		assertTrue("oldValue should be empty", logLine.getOldValue().isEmpty());
+		assertTrue("newValue shoule be empty", logLine.getNewValue().isEmpty());
+
+		// Now verify loglines:
+		assertEquals("Incorrect number of references for logline.", this.referenceChanges.size(), actualLogLine.getReferenceLogs().size());
+		final Set<Long> visitedReferences = new HashSet<Long>();
+
+		for (final ActualReferenceLog actualReferenceLog : actualLogLine.getReferenceLogs()) {
+			final Long referenceLogId = new Long(actualReferenceLog.getReferenceAttributeId());
+
+			// Have we seen this one before?
+			assertFalse("Repeated reference line " + referenceLogId, visitedReferences.contains(referenceLogId));
```

Comment:
Same as before: it's better to compare lists of names with Hamcrest

Reply:
This and the other hamcrest comments should be handled by now.